generator/sidekick/downloads_cache.go

```diff
@@ -0,0 +1,166 @@
+	}
+	// Treat `googleapis-root` as a URL to download. We want to avoid downloads
+	// if possible, so we will first try to use a cache directory in $HOME.
+	// Only if that fails we try a new download.
+	googleapisSha256, ok := rootConfig.Source["googleapis-sha256"]
+	if !ok {
+		return "", fmt.Errorf("using an https:// URL for googleapis-root requires setting googleapis-sha256")
+	}
+	cacheDir, err := downloadsCacheDir(rootConfig)
+	if err != nil {
+		return "", err
+	}
+	target := path.Join(cacheDir, googleapisSha256)
+	if isDirectory(target) {
+		return target, nil
+	}
+	tgz := target + ".tar.gz"
+	if err := downloadGoogleapisRoot(tgz, googleapisRoot, googleapisSha256); err != nil {
+		return "", err
+	}
+
+	if err := targz.Extract(tgz, cacheDir); err != nil {
+		return "", err
+	}
+	name := extractedName(rootConfig, googleapisRoot)
+	if err := os.Rename(path.Join(cacheDir, name), target); err != nil {
+		return "", err
+	}
+	return target, nil
+}
+
+func extractedName(rootConfig *Config, googleapisRoot string) string {
+	name, ok := rootConfig.Source["googleapis-extracted-name"]
+	if ok {
+		return name
+	}
+	return "googleapis-" + filepath.Base(strings.TrimSuffix(googleapisRoot, ".tar.gz"))
+}
+
+func downloadGoogleapisRoot(target, source, sha256 string) error {
+	if fileExists(target) {
+		return nil
+	}
+	var err error
+	for range 3 {
+		if err = downloadAttempt(target, source, sha256); err == nil {
+			return nil
+		}
+	}
+	return fmt.Errorf("download failed after 3 attempts, last error=%w", err)
+}
+
+func downloadAttempt(target, source, expectedSha256 string) error {
+	if err := os.MkdirAll(filepath.Dir(target), 0777); err != nil {
+		return err
+	}
+	tempFile, err := os.CreateTemp(filepath.Dir(target), "temp-")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tempFile.Name())
+
+	response, err := http.Get(source)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode >= 300 {
+		return fmt.Errorf("http error in download %s", response.Status)
+	}
+
+	if _, err := io.Copy(tempFile, response.Body); err != nil {
+		return err
+	}
+	if err := tempFile.Close(); err != nil {
+		return err
+	}
+	if err := response.Body.Close(); err != nil {
+		return err
+	}
+	file, err := os.Open(tempFile.Name())
+	if err != nil {
+		return err
+	}
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return err
+	}
+	got := fmt.Sprintf("%x", hasher.Sum(nil))
+	if expectedSha256 != got {
+		return fmt.Errorf("mismatched hash on download, expected=%s, got=%s", expectedSha256, got)
+	}
+	return os.Rename(tempFile.Name(), target)
+}
+
+func fileExists(name string) bool {
+	stat, err := os.Stat(name)
+	if err != nil {
+		return false
+	}
+	return stat.Mode().IsRegular()
+}
+
+func isDirectory(name string) bool {
+	stat, err := os.Stat(name)
+	if err != nil {
+		return false
+	}
+	if !stat.IsDir() {
+		return false
+	}
+	return true
+}
+
+func downloadsCacheDir(rootConfig *Config) (string, error) {
```

Comment:
Rename to getCacheDir? Since this function is not doing the actual download?

Reply:
Done.